### PR TITLE
Convert to Python 3

### PR DESCRIPTION
**Pull request by remcohaszing**
First the code was converted using 2to3. Then many errors were fixed
using quick fixes.

The only goal was to make Exaile run using Python 3. No cleanup was done
whatsoever.

Most things seem to work. Most notably skipping a song doesn’t.

Tests are still untouched.

Refs #246

### CODE REVIEW

**Comment by genodeftest** [comment]
Personally, I'm not opposed to this change, but we are not far from releasing Exaile 4.0 and I'd prefer to have a 4.0 release asap. In other words: I think this change should happen on a separate branch as it is changing many files and probably introducing a bunch of new bugs making the code far from a state where it can be released to a stable version.

Anyway, have you had a look at the smaller steps using `__future__` statements as proposed in #246? This way we could do incremental changes which are easier to review and will not put us so far from a working stable code base.

**Comment by virtuald** [comment]
I'm not willing to consider this until Exaile 4 is released. There are only a few issues remaining: https://github.com/exaile/exaile/milestone/2

**Comment by remcohaszing** [comment]
It makes sense to delay this until after Exaile 4 is released.

Most changes were autogenerated using `2to3`. I believe some of these changes could even be reverted.

Most remaining bugs were just about the removal of unicode encoding and decoding previously needed in Python 3.

I did not use `__future__` anywhere in this process. I didn’t need it.

I am willing to give this another try after the Exaile 4 release if you want. (This change didn’t take too long actually.) Just let me know. :smile:

File: xl/settings.py [hunk 1]
```diff
@@ -240,12 +240,16 @@ def _val_to_str(self, value):
             Turns a value of some type into a string so it
             can be a configuration value.
         """
-        for k, v in TYPE_MAPPING.iteritems():
+        # Special treatment is needed for boolean values, as bool is a
+        # subclass of int.
+        if isinstance(value, bool):
+            return 'B: ' + str(value)
```
Comment (remcohaszing) [inline xl/settings.py]: This change actually fixes a bug which is consistent in Python 3, but still may happen randomly in Python 2.

Comment (sjohannes) [reply xl/settings.py]: Thanks; I've committed this fix as e677d2fc.

**Comment by genodeftest** [comment]
Is it ok if I close this PR in favor of merging the [py3](https://github.com/exaile/exaile/tree/py3) branch into master (hopefully, some day)?

**Comment by sjohannes** [comment]
Sorry that this PR didn't end up getting used. For our Python 3 port I wanted to base it on `futurize` rather than `2to3` as that seems to be the more developed converter, so that's what we used in the py3 branch. Rok and I have been testing the branch and fixing a lot of the subtle issues so we'll be going with that.

**Comment by raveit65** [comment]
Looks like `Exaile 4.0.2 from 14 Jan` doesn't include the python3 port.
Can we expect a new release?
I'd like to include Exaile again for Mate-Compiz spin from fedora. We're are already in beta freeze for fedora 32, and it take some time to rebuild and get a new exaile version in repos before we reach release freeze.
Thank you

**Comment by sjohannes** [comment]
If you're okay with the current state of the master branch, we can tag an alpha release so you can start working on the packaging. We can then do another alpha/beta release when you need to submit the final version.

I can't guarantee everything will be working by that time, but AFAIK only a few plugins are broken right now (@genodeftest would know more about this) so it shouldn't be too bad I think.

**Comment by raveit65** [comment]
Np, when there are a few minimal issues. Main problem is the python2 is dropped from fedora 32, that means older versions can't be build anymore. An alpha release should make sure than no one has the idea to retire Exaile from fedora ;)
I will notice fedoras Exaile maintainer about a new release when it is done.
Can you ping me here?
Thanks.

**Comment by genodeftest** [comment]
I'm fine with pushing the python3 stuff out soon. I have an open PR for the `cd` plugin which needs to be rebased to the current master but otherwise the plugins should still work. IMHO we can do an alpha.

@raveit65: apparently, I'm the new exaile maintainer for Fedora but if you want to help, please feel free to do so.

**Comment by raveit65** [comment]
@genodeftest 
I am happy to test an alpha release at bodhi, please ping me.

**Pull request closed without merging**